sys/tools/syscalls/core/syscall.lua

```diff
@@ -0,0 +1,444 @@
+function syscall:add(line)
+    if self:addDef(line) then
+        -- Cases where the syscalls.master entry is one line; we just want to
+        -- exit and add:
+        if self.name ~= "{" then
+            -- A NIL name should be written as an empty string.
+            if self.name == nil then
+                self.name = ""
+            end
+            self.alias = self.name -- set for all cases
+
+            -- This system call is a range.
+            if tonumber(self.num) == nil then
+                return true
+            -- This system call is a loadable system call.
+            elseif self.altname ~= nil and self.alttag ~= nil and
+                   self.rettype ~= nil then
+                self.cap = "0"
+                self.thr = "SY_THR_ABSENT"
+                self.arg_alias = self:symbol() .. "_args"
+                self.ret = self.rettype
+                return true
+            -- This system call is some other one line entry.
+            else
+                return true
+            end
+        end
+        return false -- Otherwise, definition added; keep going.
+    end
+    if self:addFunc(line) then
+        return false -- Function added; keep going.
+    end
+    if self:addArgs(line) then
+        return false -- Arguments added; keep going.
+    end
+    return self:isAdded(line) -- Final validation, before adding.
+end
+
+-- Return TRUE if this system call is native.
+-- NOTE: The other system call names are also treated as native, so that's why
+-- they're being allowed in here.
+function syscall:native()
+    return self:compat_level() == native or self.name == "lkmnosys" or
+           self.name == "sysarch"
+end
+
+function syscall:new(obj)
+	obj = obj or { }
+	setmetatable(obj, self)
+	self.__index = self
+
+	self.expect_rbrace = false
+    self.changes_abi = false
+	self.args = {}
+
+	return obj
+end
+
+-- Make a shallow copy of `self` and replace the system call number with num
+-- (which should be a number).
+-- For system call ranges.
+function syscall:shallowCopy(num)
+	local obj = syscall:new()
+
+	-- shallow copy
+	for k, v in pairs(self) do
+		obj[k] = v
+	end
+	obj.num = num	-- except override range
+	return obj
+end
+
+-- Make a deep copy of the parameter object.
+-- CREDIT: http://lua-users.org/wiki/CopyTable
+-- For a full system call (the nested arguments table should be a deep copy).
+local function deepCopy(orig)
+    local type = type(orig)
+    local copy
+
+    if type == 'table' then
+        copy = {}
+        for orig_key, orig_value in next, orig, nil do
+            copy[deepCopy(orig_key)] = deepCopy(orig_value)
+        end
+        setmetatable(copy, deepCopy(getmetatable(orig)))
+    else -- number, string, boolean, etc
+        copy = orig
+    end
+
+    return copy
+end
+
+-- CREDIT: http://lua-users.org/wiki/CopyTable
+-- Save copied tables in `copies`, indexed by original table.
+function deepCopy(orig, copies)
```

Comment:
It doesn't seem right to have two slightly different implementations of deepCopy()